src/term/check/exhaustiveness.rs

```diff
@@ -0,0 +1,463 @@
+    typ.extend_from_slice(&problem.types[1 ..]);
+
+    // specialize the cases for the current constructor
+    let cases = specialize(&problem.case, constructor, constructor_type.len()).concat();
+
+    let mut problem_ = Problem { matrix, case: cases, types: typ };
+
+    useful(self, &mut problem_)
+  }
+}
+
+// if a signature is complete, we split and specialize the matrix
+// into new problems for each constructor of a type
+fn split(ctx: &Ctx, typ: &Name, args: Vec<RulePat>, problem: &Problem) -> bool {
+  if let Some(typ) = ctx.ctx_types.get(&typ.0) {
+    typ.iter().any(|label| ctx.specialize_ctr(&Name(label.clone()), args.clone(), problem))
+  } else {
+    false
+  }
+}
+
+fn is_wildcard(pat: &RulePat) -> bool {
+  matches!(pat, RulePat::Var(..))
+}
+
+fn is_wildcard_row(row: &Row) -> bool {
+  row.first().is_some_and(is_wildcard)
+}
+
+fn is_wildcard_matrix(matrix: &Matrix) -> bool {
+  matrix.iter().all(is_wildcard_row)
+}
+
+fn useful(ctx: &Ctx, problem: &mut Problem) -> bool {
+  match problem {
+    // if the matrix has 0 columns and 0 rows the case is useful
+    // to the matrix
+    problem if problem.is_empty() => true,
+
+    // if the matrix has 0 columns but more than 0 rows the case is useless
+    // to the matrix
+    problem if problem.is_complete() => false,
+
+    problem => {
+      // take the types of the patterns
+      if let Some((name, args)) = problem.type_name() {
+        match &problem.case[0] {
+          // if it is a wildcard/var
+          RulePat::Var(..) => {
+            if is_wildcard_matrix(&problem.matrix) {
+              // if all patterns of the first column are wildcards then
+              // get the default matrix and check again
+              useful(ctx, &mut default_matrix(problem))
+            } else {
+              let mut cons: Option<Vec<String>> =
+                problem.matrix.iter().map(|row| get_pat_ctr_name(row[0].clone())).collect();
+              if cons.is_none() {
+                cons = Some(vec![]);
+              }
+              if let Some(names) = cons {
+                match is_sig_complete(ctx, name.0.clone(), names) {
+                  // if the signature is complete then split based on the type
+                  Completeness::Complete(_) => split(ctx, &name, args, problem),
+                  // if the signature of the first pattern is incomplete then get the default matrix
+                  // and check again
+                  Completeness::Incomplete(_) => useful(ctx, &mut default_matrix(problem)),
+                }
+              } else {
+                false
+              }
+            }
+          }
+          // if it is a constructor we specialize this constructor case
+          RulePat::Ctr(nam_, ..) => ctx.specialize_ctr(&nam_, args, &problem),
+        }
+      } else {
+        false
+      }
+    }
+  }
+}
+
+impl Book {
+  pub fn check_exhaustiveness(&self) -> anyhow::Result<()> {
```

Comment:
Could you add a general description of how this algorithm works? I already forgot the details and I'm failing to understand it now